twister: No skips for integration platforms in integration mode

PerMac opened the pull request:
All skips for platforms listed in the .yaml's integration_platforms
will be treated as errors in the integration mode ('-G/--integration').
This feature serves to guarantee a proper testing scope in CI as no
integration platform will be skipped silently.
Fixes #33874

Signed-off-by: Maciej Perkowski <Maciej.Perkowski@nordicsemi.no>

PerMac committed:
twister: No skips for integration platforms in integration mode

All skips for platforms listed in the .yaml's integration_platforms
will be treated as errors in the integration mode ('-G/--integration').
This feature serves to guarantee a proper testing scope in CI as no
integration platform will be skipped silently.
Fixes #33874

Signed-off-by: Maciej Perkowski <Maciej.Perkowski@nordicsemi.no>

PerMac commented:
This feature can be tested e.g. by changing "ram: 256" to "ram: 1" in zephyr/boards/arm/nrf52840dk_nrf52840/nrf52840dk_nrf52840.yaml
and calling `./scripts/twister -T ./tests/drivers/flash/ -G` with and without extra  `--no-integration-skips`

PerMac commented:
It looks like the compliance fail has nothing to do with my PR.

galak reviewed (dismissed):
I was looking at something similar based on my other PR w/regards to platform_allow.  I think we should just treat this as an error in general and not add a command line option for it.

However, I think there's a subtle issue here with regards to the default integration_platforms not supporting all tests, so wouldn't we hit this all the time?

In general I'm thinking that with `-G` that if a test is "selected" we should treat it as an error if there isn't at least one platform the test is built on.

nashif reviewed (changes requested):
agree with @galak. This PR seems to be introducing a micro enhancement covering only a small usecase (first you have to be in -G mode, then you have to enable yet another option). 

If our goal is to have minimal coverage (at least 1) for each test when using the default twister options, i.e. calling `twister -T <test>` then it should be globally enabled and twister should report error if nothing is run.

default platforms are basically the default `integration_platforms` that can be overloaded with a specific set, so the same rules should apply.

galak commented:
In some local work I added:
```
diff --git a/scripts/pylib/twister/twisterlib.py b/scripts/pylib/twister/twisterlib.py
index 38a5f2615d..acf6397e4b 100755
--- a/scripts/pylib/twister/twisterlib.py
+++ b/scripts/pylib/twister/twisterlib.py
@@ -3175,6 +3175,7 @@ class TestSuite(DisablePyTestCollectionMixin):
                         aa = list(filter(lambda tc: tc.platform.name in c, instance_list))
                         self.add_instances(aa)
                     else:
+                        print("ERROR: platform_allow and no integration overlap %s" % tc.name)
                         self.add_instances(instance_list[:1])
                 else:
                     instances = list(filter(lambda tc: tc.platform.default, instance_list))
```

As at least one place we should possibly treat as an error with `-G` set.  As the `self.add_instances` doesn't accomplish anything.

PerMac commented:
@galak I think there still some misunderstanding between us about the idea behind integration_platforms, platform_allow, and default platforms. This is how I see this:
> I was looking at something similar based on my other PR w/regards to platform_allow. I think we should just treat this as an error in general and not add a command line option for it.

I agree and can make this a part of `-G` without the need for extra arg. But this is only for integration mode (CI runs on-commit) and platforms on the integration_platforms list.

> However, I think there's a subtle issue here with regards to the default integration_platforms not supporting all tests, so wouldn't we hit this all the time?

There is no "default integration_platforms" per se, only a suggestion on which platform to use there. The idea behind the integration_platforms is that this list is defined individually on the test level in test's .yaml. In contrast, there are "default platforms for testing", which are all platforms that have "testing: default: true: in their board.yamls and they have "global" scope. If twister is called without explicitly defining which platforms to use, those are set as the platforms to use. In addition, if we use integration mode (`-G`) and test's yaml defines integration_platforms, the platforms from this particular integration_platforms list overwrite the default platforms for that particular test. This is what happens in the on-commit CI.

> In general I'm thinking that with -G that if a test is "selected" we should treat it as an error if there isn't at least one platform the test is built on.

I think this is a good idea. However, I think this should be solved in a new PR. Will you create an issue for this?

PerMac commented:
@nashif 
> This PR seems to be introducing a micro enhancement covering only a small usecase (first you have to be in -G mode, then you have to enable yet another option).

The change is small indeed but I think it has some greater consequences. I also think that -G mode is not a small usecase as on-commits regression checks are a rather major part of testing and often the only place where developers have contact with tests. 

One of the really common issues is not having the proper coverage in on-commit tests. I am constantly seeing developers nagging that due to the scope issue (test was skipped for whatever reason) the CI test didn't found an error which popped out only later on. This is why I wanted to put some greater stress on the integration_platforms usage and its functionality. 

I will set this check as part of `-G` so there will be no need for extra arg. This is what I meant when I asked you if this should be the default check.

> If our goal is to have minimal coverage (at least 1) for each test when using the default twister options, i.e. calling twister -T <test> 

Not exactly. The goal is to make sure that if 'integration_platforms' are listed in a test, the test in the CI will get built/executed on each of them. No skips allowed on these platforms with `-G`.

>then it should be globally enabled and twister should report an error if nothing is run.

This can be used as some general condition. I think Kumar's idea and my enhancement request relate to `-G` option, but it can be generalized to all use cases as you suggest. 

> default platforms are basically the default integration_platforms that can be overloaded with a specific set, so the same rules should apply.

I see this in a slightly different way, hence my PR.  I opt that with `-G' the platforms on the integration_platforms list are treated in a special way if a test defines them. Such a test will have to pass on all of its integration platforms (not just on any single) and no skips are allowed.

PerMac commented:
@galak In regards to your comment starting with `In some local work I added:`. Do you have some tests in mind where integration_platforms and platforms_allow have no common part? We could then try this PR on it. If I did this right we would get an error: "Not in testcase platform allow list but is one of the integration platforms"

nashif commented:
> I see this in a slightly different way, hence my PR. I opt that with `-G' the platforms on the integration_platforms list are treated in a special way if a test defines them. Such a test will have to pass on all of its integration platforms (not just on any single) and no skips are allowed.

I am fine with `integration_platforms` become a no-skip case, however, the rationale for `integration_platforms needs to stay intact, i.e. it should be a minimal list from a coverage perspective and not `a list of everything that I want to always run because this is my platform and I do not care about anything else`

PerMac commented:
@nashif @galak I've deleted the need for extra arg. Now error when skipping integration platforms is part of integration mode (-G). I modified the title and the commit msg accordingly.
